internal/services/v1/debug.go

```diff
@@ -0,0 +1,188 @@
+package v1
+
+import (
+	"context"
+	"strings"
+
+	v1 "github.com/authzed/authzed-go/proto/authzed/api/v1"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	cexpr "github.com/authzed/spicedb/internal/caveats"
+	"github.com/authzed/spicedb/pkg/datastore"
+	dispatch "github.com/authzed/spicedb/pkg/proto/dispatch/v1"
+	"github.com/authzed/spicedb/pkg/schemadsl/compiler"
+	"github.com/authzed/spicedb/pkg/schemadsl/generator"
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+// convertDispatchDebugInformation converts dispatch debug information found in the response metadata
+// into DebugInformation returnable to the API.
+func convertDispatchDebugInformation(
+	ctx context.Context,
+	caveatContext map[string]any,
+	metadata *dispatch.ResponseMeta,
+	reader datastore.Reader,
+) (*v1.DebugInformation, error) {
+	debugInfo := metadata.DebugInfo
+	if debugInfo == nil {
+		return nil, nil
+	}
+
+	caveats, err := reader.ListCaveats(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	namespaces, err := reader.ListNamespaces(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defs := make([]compiler.SchemaDefinition, 0, len(namespaces)+len(caveats))
+	for _, caveat := range caveats {
+		defs = append(defs, caveat)
+	}
+	for _, ns := range namespaces {
+		defs = append(defs, ns)
+	}
+
+	schema, _ := generator.GenerateSchema(defs)
+	converted, err := convertCheckTrace(ctx, caveatContext, debugInfo.Check, reader)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.DebugInformation{
+		Check:      converted[0],
+		SchemaUsed: strings.TrimSpace(schema),
+	}, nil
+}
+
+func convertCheckTrace(ctx context.Context, caveatContext map[string]any, ct *dispatch.CheckDebugTrace, reader datastore.Reader) ([]*v1.CheckDebugTrace, error) {
+	traces := make([]*v1.CheckDebugTrace, 0, len(ct.Request.ResourceIds))
+	for _, resourceID := range ct.Request.ResourceIds {
+		permissionType := v1.CheckDebugTrace_PERMISSION_TYPE_UNSPECIFIED
+		if ct.ResourceRelationType == dispatch.CheckDebugTrace_PERMISSION {
+			permissionType = v1.CheckDebugTrace_PERMISSION_TYPE_PERMISSION
+		} else if ct.ResourceRelationType == dispatch.CheckDebugTrace_RELATION {
+			permissionType = v1.CheckDebugTrace_PERMISSION_TYPE_RELATION
+		}
+
+		subRelation := ct.Request.Subject.Relation
+		if subRelation == tuple.Ellipsis {
+			subRelation = ""
+		}
+
+		permissionship := v1.CheckDebugTrace_PERMISSIONSHIP_NO_PERMISSION
+		checkResult, ok := ct.Results[resourceID]
+		if ok {
+			if checkResult.Membership == dispatch.ResourceCheckResult_MEMBER {
+				permissionship = v1.CheckDebugTrace_PERMISSIONSHIP_HAS_PERMISSION
```

Comment:
The test does not exercise this codepath. I imagine a situation where, even though caveats were part of the graph, they where short-circuited. In that case `checkResult.Expression` would be `nil`. Perhaps could use a test.

Reply:
Which part isn't exercised?

Reply:
`CheckDebugTrace_PERMISSIONSHIP_HAS_PERMISSION`